File: solutions/typescript/021-merge-k-sorted-lists.ts
/**
 * Problem 21: Merge k Sorted Lists (LeetCode 23)
 * Difficulty: Hard
 * Language: TypeScript
 */
function mergeKLists(lists: (ListNode | null)[]): ListNode | null {
  const merge = (
    l1: ListNode | null,
    l2: ListNode | null
  ): ListNode | null => {
    if (!l1) return l2;
    if (!l2) return l1;
    if (l1.val <= l2.val) return new ListNode(l1.val, merge(l1.next, l2));
    return new ListNode(l2.val, merge(l1, l2.next));
  };

  return lists.reduce<ListNode | null>(
    (acc, list) => merge(acc, list),
    null
  );
}

File: solutions/rust/021-merge-k-sorted-lists.rs
/**
 * Problem 21: Merge k Sorted Lists (LeetCode 23)
 * Difficulty: Hard
 * Language: Rust
 */
pub fn merge_k_lists(lists: Vec<Option<Box<ListNode>>>) -> Option<Box<ListNode>> {
    fn merge(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        match (l1, l2) {
            (None, r) => r,
            (l, None) => l,
            (Some(a), Some(b)) if a.val <= b.val =>
                Some(Box::new(ListNode { val: a.val, next: merge(a.next, Some(b)) })),
            (Some(a), Some(b)) =>
                Some(Box::new(ListNode { val: b.val, next: merge(Some(a), b.next) })),
        }
    }
    lists.into_iter().fold(None, |acc, list| merge(acc, list))
}
